File: src/data/moex.py
"""MOEX ISS API data fetcher with automatic demo fallback.

Tries the live MOEX ISS API first. If the network is unavailable,
transparently falls back to bundled demo data.
"""

import datetime as dt
import os
from typing import Optional

import pandas as pd
import requests

from src.config import MOEX_ISS_BASE

_SESSION = requests.Session()
_SESSION.headers.update({
    "User-Agent": "Mozilla/5.0 (compatible; MOEXDashboard/1.0)",
})

_DEMO_DIR = os.path.join(os.path.dirname(__file__), "..", "..", "demo_data")
_USE_DEMO = None  # auto-detect on first call


def _is_api_available() -> bool:
    global _USE_DEMO
    if _USE_DEMO is not None:
        return not _USE_DEMO
    try:
        r = _SESSION.get(f"{MOEX_ISS_BASE}/index.json", timeout=5)
        available = r.status_code == 200
    except Exception:
        available = False
    _USE_DEMO = not available
    if _USE_DEMO:
        print("  [INFO] MOEX API unavailable — using demo data")
    else:
        print("  [INFO] Connected to MOEX ISS API")
    return available


def _load_demo_candles(ticker: str) -> pd.DataFrame:
    path = os.path.join(_DEMO_DIR, f"{ticker}.csv")
    if not os.path.exists(path):
        return pd.DataFrame(columns=["datetime", "open", "high", "low", "close", "volume"])
    df = pd.read_csv(path, parse_dates=["datetime"])
    return df


def _load_demo_prices(tickers: list[str]) -> pd.DataFrame:
    path = os.path.join(_DEMO_DIR, "_prices.csv")
    if not os.path.exists(path):
        return pd.DataFrame(columns=["ticker", "last", "change_pct", "volume", "time"])
    df = pd.read_csv(path)
    return df[df["ticker"].isin(tickers)].reset_index(drop=True)


def fetch_candles(
    ticker: str,
    interval: int = 60,
    start: Optional[str] = None,
    end: Optional[str] = None,
) -> pd.DataFrame:
    """Fetch OHLCV candles from MOEX ISS (or demo data if unavailable)."""
    if not _is_api_available():
        return _load_demo_candles(ticker)

    if start is None:
        start = (dt.date.today() - dt.timedelta(days=180)).isoformat()
    if end is None:
        end = dt.date.today().isoformat()

    url = f"{MOEX_ISS_BASE}/engines/stock/markets/shares/securities/{ticker}/candles.json"

    all_rows = []
    cursor_start = 0

    while True:
        params = {
            "from": start,
            "till": end,
            "interval": interval,
            "start": cursor_start,
        }
        resp = _SESSION.get(url, params=params, timeout=15)
        resp.raise_for_status()
        data = resp.json()

        candles = data.get("candles", {})
        columns = candles.get("columns", [])
        rows = candles.get("data", [])

        if not rows:
            break

        all_rows.extend(rows)
        cursor_start += len(rows)

        if len(rows) < 500:
            break

    if not all_rows:
        return pd.DataFrame(columns=["datetime", "open", "high", "low", "close", "volume"])

    df = pd.DataFrame(all_rows, columns=columns)

    rename_map = {
        "open": "open",
        "close": "close",
        "high": "high",
        "low": "low",
        "value": "value",
        "volume": "volume",
        "begin": "datetime",
        "end": "end_dt",
    }
    df = df.rename(columns=rename_map)
    df["datetime"] = pd.to_datetime(df["datetime"])
    df = df.sort_values("datetime").reset_index(drop=True)

    return df[["datetime", "open", "high", "low", "close", "volume"]]


def fetch_current_prices(tickers: list[str]) -> pd.DataFrame:
    """Fetch latest market data for multiple tickers."""
    if not _is_api_available():
        return _load_demo_prices(tickers)

    url = f"{MOEX_ISS_BASE}/engines/stock/markets/shares/securities.json"
    params = {
        "iss.meta": "off",
        "securities.columns": "SECID,LAST,LASTTOPREVPRICE,VOLTODAY,TIME",
    }
    resp = _SESSION.get(url, params=params, timeout=15)
    resp.raise_for_status()
    data = resp.json()

    securities = data.get("securities", {})
    columns = securities.get("columns", [])
    rows = securities.get("data", [])

    df = pd.DataFrame(rows, columns=columns)
    df = df.rename(columns={
        "SECID": "ticker",
        "LAST": "last",
        "LASTTOPREVPRICE": "change_pct",
        "VOLTODAY": "volume",
        "TIME": "time",
    })

    df = df[df["ticker"].isin(tickers)]
    return df.reset_index(drop=True)


def get_available_tickers(limit: int = 50) -> list[str]:
    """Get a list of actively traded share tickers on MOEX."""
    if not _is_api_available():
        return [f.replace(".csv", "") for f in os.listdir(_DEMO_DIR)
                if f.endswith(".csv") and not f.startswith("_")][:limit]

    url = f"{MOEX_ISS_BASE}/engines/stock/markets/shares/securities.json"
    params = {
        "iss.meta": "off",
        "marketdata.columns": "SECID,VOLTODAY",
    }
    resp = _SESSION.get(url, params=params, timeout=15)
    resp.raise_for_status()
    data = resp.json()

    md = data.get("marketdata", {})
    columns = md.get("columns", [])
    rows = md.get("data", [])

    df = pd.DataFrame(rows, columns=columns)
    df = df[df["VOLTODAY"] > 0]
    df = df.sort_values("VOLTODAY", ascending=False)

    return df["SECID"].head(limit).tolist()

File: js/moex.js
const FETCH_TIMEOUT = 10000;
const MAX_RETRIES = 2;

async function fetchWithRetry(url) {
    for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
        const ctrl = new AbortController();
        const timer = setTimeout(() => ctrl.abort(), FETCH_TIMEOUT);
        try {
            const resp = await fetch(url, { signal: ctrl.signal });
            clearTimeout(timer);
            if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
            return await resp.json();
        } catch (e) {
            clearTimeout(timer);
            if (attempt === MAX_RETRIES) throw e;
            await new Promise(r => setTimeout(r, 1000 * (attempt + 1)));
        }
    }
}

function parseMoscowTime(dateStr) {
    if (!dateStr) return 0;
    const clean = dateStr.replace(' ', 'T');
    if (/[+-]\d{2}:?\d{2}$/.test(clean) || clean.endsWith('Z')) return Math.floor(new Date(clean).getTime() / 1000);
    return Math.floor(new Date(clean + '+03:00').getTime() / 1000);
}

export async function fetchCandles(ticker, interval, days) {
    const end = new Date().toISOString().slice(0, 10);
    const start = new Date(Date.now() - days * 86400000).toISOString().slice(0, 10);
    const base = 'https://iss.moex.com/iss/engines/stock/markets/shares/securities';
    let allRows = [], cursor = 0;

    while (true) {
        const url = `${base}/${ticker}/candles.json?from=${start}&till=${end}&interval=${interval}&start=${cursor}&iss.meta=off`;
        const data = await fetchWithRetry(url);
        const rows = data.candles.data;
        if (!rows.length) break;
        allRows = allRows.concat(rows);
        cursor += rows.length;
        if (rows.length < 500) break;
    }

    return allRows.map(r => ({
        time: interval === 24 || interval === 7 || interval === 31
            ? r[6].slice(0, 10)
            : parseMoscowTime(r[6]),
        open: r[0], close: r[1], high: r[2], low: r[3], volume: r[5],
    }));
}

export async function fetchPrices(tickers) {
    const url = `https://iss.moex.com/iss/engines/stock/markets/shares/securities.json?securities=${tickers.join(',')}&iss.meta=off&iss.only=securities,marketdata`;
    const data = await fetchWithRetry(url);
    const secCols = data.securities.columns;
    const mktCols = data.marketdata.columns;
    const ci = (cols, name) => cols.indexOf(name);
    const secMap = {};
    for (const r of data.securities.data) {
        const id = r[ci(secCols, 'SECID')];
        if (!secMap[id]) secMap[id] = { shortname: r[ci(secCols, 'SHORTNAME')] };
    }
    const result = [];
    const seen = new Set();
    for (const r of data.marketdata.data) {
        const id = r[ci(mktCols, 'SECID')];
        if (seen.has(id)) continue;
        const last = r[ci(mktCols, 'LAST')];
        if (last == null) continue;
        seen.add(id);
        result.push({
            ticker: id,
            shortname: secMap[id]?.shortname || id,
            last, change: r[ci(mktCols, 'LASTCHANGEPCT')] ?? null,
            volume: r[ci(mktCols, 'VOLTODAY')],
            time: r[ci(mktCols, 'UPDATETIME')] || '',
        });
    }
    return result;
}

export async function fetchOrderBook(ticker) {
    const url = `https://iss.moex.com/iss/engines/stock/markets/shares/boards/TQBR/securities/${ticker}/orderbook.json?iss.meta=off`;
    const data = await fetchWithRetry(url);
    const cols = data.orderbook.columns;
    const ci = (name) => cols.indexOf(name);
    const bids = [], asks = [];
    for (const row of data.orderbook.data) {
        const entry = { price: row[ci('PRICE')], quantity: row[ci('QUANTITY')] };
        if (row[ci('BUYSELL')] === 'B') bids.push(entry);
        else asks.push(entry);
    }
    bids.sort((a, b) => b.price - a.price);
    asks.sort((a, b) => a.price - b.price);
    return { bids: bids.slice(0, 15), asks: asks.slice(0, 15) };
}
